src/viam/sdk/components/encoder/encoder.hpp

```diff
@@ -0,0 +1,114 @@
+/// @file components/encoder/encoder.hpp
+///
+/// @brief Defines a `Encoder` component.
+#pragma once
+
+#include <string>
+
+#include <viam/api/component/encoder/v1/encoder.pb.h>
+
+#include <viam/sdk/common/proto_type.hpp>
+#include <viam/sdk/common/utils.hpp>
+#include <viam/sdk/config/resource.hpp>
+#include <viam/sdk/registry/registry.hpp>
+#include <viam/sdk/resource/resource_manager.hpp>
+
+namespace viam {
+namespace sdk {
+
+/// @defgroup Encoder Classes related to the `Encoder` component.
+
+/// @class EncoderSubtype
+/// @brief Defines a `ResourceSubtype` for the `Encoder` component.
+/// @ingroup Encoder
+class EncoderSubtype : public ResourceSubtype {
+   public:
+    std::shared_ptr<ResourceServerBase> create_resource_server(
+        std::shared_ptr<ResourceManager> manager) override;
+    std::shared_ptr<ResourceBase> create_rpc_client(std::string name,
+                                                    std::shared_ptr<grpc::Channel> chan) override;
+    EncoderSubtype(const google::protobuf::ServiceDescriptor* service_descriptor)
+        : ResourceSubtype(service_descriptor){};
+};
+
+/// @class Encoder encoder.hpp "components/encoder/encoder.hpp"
+/// @brief An encoder is a device that is hooked up to motors to report a position
+/// @ingroup Encoder
+///
+/// This acts as an abstract base class to be inherited from by any drivers representing
+/// specific encoder implementations. This class cannot be used on its own.
+class Encoder : public ComponentBase {
+   public:
+    /// @enum position_type
+    enum position_type {
```

Comment:
This is an un-scoped enum, so its names leak into the enclosing scope, in this case the `Encoder` class. So that means that `Encoder::ANGLE_DEGREES` is a meaningful symbol. There is nothing wrong with that if that is what you want, but it might be good to put `POSITION_TYPE_`  in before each value if you want to do that:

```
class Encoder ... {
public:
    enum position_type {
        POSITION_TYPE_TICKS_COUNT,
        POSITION_TYPE_ANGLE_DEGREES,
    };
```

That way `Encoder::POSITION_TYPE_ANGLE_DEGREES` is clearly a position type.

Alternatively, you can make it an `enum class`, and then the prefix is the actual typename of the `enum` and the values don't leak into the surrounding scope:

```
class Encoder ... {
public:
    enum class position_type {
        TICKS_COUNT,
        ANGLE_DEGREES
    }
```

Means you get `Encoder::position_type::TICKS_COUNT`.

So I recommend either going to an `enum class` or prefixing the enumeration values.

Reply:
Ahh, I hadn't used `enum class` before. Will switch. Thank you.